src/Orleans.Persistence.CosmosDB/Options/IPartitionKeyProvider.cs

```diff
@@ -0,0 +1,20 @@
+using Orleans.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Orleans.Persistence.CosmosDB.Options
+{
+    public interface IPartitionKeyProvider
+    {
+        string GetPartitionKey(string grainType, GrainReference grainReference);
```

Comment:
I would make this `Task<string>`. That would allow people to even go further and request the partition key map from external systems. 

The default would be just `Task.FromResult(grainType)` anyway.